### PR TITLE
Update GIR include paths

### PR DESCRIPTION
**Pull request by QuLogic**
This modifies the search paths for GIR generation:

* Include paths coming from the target only add its source path. Instead add include paths in a manner similar to the ninja backend, with both source and build directories.
* Add include paths from all types of `Dependency`s of the target _and_ pull those dependencies directly from the target from which the GIR is being built.
* Previously, to "link" with another GIR being built, you had to add the name in the `includes` argument and the `GirTarget` in the `dependencies` argument to set the search path. Now, you can specify the `GirTarget` in the `includes` and the correct name & search path are added.
* Confusingly, the `include_directories` applied to the search path for GIR files only. Now it's applied to header searches as well.

I probably need to come up with some way to test this...

### CODE REVIEW

**Comment by QuLogic** [comment]
Hmm, that last commit might be causing some warnings.

File: mesonbuild/modules/gnome.py [hunk 1]
```diff
@@ -188,41 +192,47 @@ def generate_gir(self, state, args, kwargs):
...
-                            scan_command += [lib]
-
-                    girdir = dep.held_object.get_variable ("girdir")
-                    if girdir:
-                        scan_command += ["--add-include-path=%s" % girdir]
-                else:
-                    mlog.log('dependency %s not handled to build gir files' % dep)
-                    continue
+        deps = (girtarget.get_all_link_deps() + girtarget.get_external_deps() +
+                kwargs.pop('dependencies', []))
```
Comment (thiblahute) [inline mesonbuild/modules/gnome.py]: You are bringing in the list of dependencies the `SharedLibrary` which is not handled in the following loop.

File: mesonbuild/modules/gnome.py [hunk 2]
```diff
@@ -188,41 +192,47 @@ def generate_gir(self, state, args, kwargs):
...
+                        scan_command += ["-L%s" % os.path.dirname(lib)]
+                        libname = os.path.basename(lib)
+                        if libname.startswith("lib"):
+                            libname = libname[3:]
+                        libname = libname.split(".so")[0]
+                        lib = "-l%s" % libname
+                    # Hack to avoid passing some compiler options in
+                    if lib.startswith("-W"):
+                        continue
+                        scan_command += [lib]
```
Comment (thiblahute) [inline mesonbuild/modules/gnome.py]: Hrm, that line is never reached, probably not your fault though.

File: mesonbuild/modules/gnome.py (hunk 1 above)
Comment (QuLogic) [inline mesonbuild/modules/gnome.py]: Hmm yes, should I just ignore them, or not add them at all?

Comment (thiblahute) [inline mesonbuild/modules/gnome.py]: Why did you add them in the first place? :)

File: mesonbuild/modules/gnome.py (hunk 2 above)
Comment (QuLogic) [inline mesonbuild/modules/gnome.py]: You're right; that's weird, but that's the way it was before. Maybe it's not supposed to be indented this much?

File: mesonbuild/modules/gnome.py (hunk 1 above)
Comment (QuLogic) [inline mesonbuild/modules/gnome.py]: Well, I added external deps because I want the include paths to match the target library's. I thought `get_all_link_deps` would be covered by `InternalDependency`, but I guess not entirely.

File: mesonbuild/modules/gnome.py [hunk 3]
```diff
@@ -188,54 +206,65 @@ def generate_gir(self, state, args, kwargs):
...
-                    continue
-
-        inc_dirs = None
-        if kwargs.get('include_directories'):
-            inc_dirs = kwargs.pop('include_directories')
-
-            if not isinstance(inc_dirs, list):
-                inc_dirs = [inc_dirs]
+        deps = (girtarget.get_all_link_deps() + girtarget.get_external_deps() +
+                kwargs.pop('dependencies', []))
```
Comment (thiblahute) [inline mesonbuild/modules/gnome.py]: Maybe just add support for those in the following loop?

**Comment by QuLogic** [comment]
I think I have this working the way I want now; I've updated the original post to summarize the changes. Now I just need to write some tests to demonstrate the changes.

**Comment by QuLogic** [comment]
I rebased and added some tests; I'm not sure if they exercise everything, but the build appears to work for my purposes.

**Comment by QuLogic** [comment]
PS, please update the wiki for `gnome.generate_gir`. It currently states that nothing is returned, but that's not true; it returns two targets, one for the `.gir` and one for the `.typelib`.

File: mesonbuild/modules/gnome.py [hunk 4]
```diff
@@ -188,54 +213,66 @@ def generate_gir(self, state, args, kwargs):
...
-                            libname = libname.split(".so")[0]
-                            lib = "-l%s" % libname
-                        # Hack to avoid passing some compiler options in
-                        if lib.startswith("-W"):
-                            continue
-                            scan_command += [lib]
-
-                    girdir = dep.held_object.get_variable ("girdir")
+        deps = (girtarget.get_all_link_deps() + girtarget.get_external_deps() +
+                kwargs.pop('dependencies', []))
```
Comment (jpakkane) [inline mesonbuild/modules/gnome.py]: `dependencies` might not be an array which makes this operation fail. Pop it first and wrap in `[]` if necessary.

File: mesonbuild/modules/gnome.py (hunk 2 above)
Comment (jpakkane) [inline mesonbuild/modules/gnome.py]: Indent it one less and if nothing breaks let's call it correct. :)

Comment (QuLogic) [inline mesonbuild/modules/gnome.py]: Seems to pass tests and work for me.

File: mesonbuild/modules/gnome.py (hunk 4 above)
Comment (QuLogic) [inline mesonbuild/modules/gnome.py]: Fixed.

**Comment by thiblahute** [comment]
What are we waiting for that one? I want to work on fixing the gtkdoc generation (it is full of bugs) but it will conflict a lot with that.

**Comment by jpakkane** [comment]
I wanted to focus on the mkenums MR first (because at the time it looked like to be very close to merging) before this one. This got delayed by that and now by me being at LAS. I'll try to work through the backlog during the weekend.
